frontend.php

```diff
@@ -0,0 +1,198 @@
+<?php
+
+
+/*
+ * Outputs the HTML required to display the text in the about me section.
+ *
+ * @param PDO $db The database object to retrieve the about me information from.
+ *
+ * @return string Returns the Name, Title and Description wrapped in the correct html tags.
+ */
+function displayAboutMe(PDO $db) {
+    $stmt = $db->prepare('SELECT `name`,`title`,`desc` FROM `about`');
+    $stmt->execute();
+    $result = $stmt->fetch();
+    $name = '<h1>' . $result['name'] . '</h1>';
+    $title = '<h2>' . $result['title'] . '</h2>';
+    $desc = '<p>' . $result['desc'] . '</p>';
+    return $name . $title . $desc;
+}
+
+/*
+ * Outputs the HTML required to display the badges in the about me section.
+ *
+ * @param PDO $db The database object to retrieve the about me information from.
+ *
+ * @return string Returns each badge in the database, wrapped in an <i> tag.
+ */
+function displayBadges(PDO $db) {
+    $stmt = $db->prepare('SELECT `value` FROM `badges`');
+    $stmt->execute();
+    $entries = $stmt->fetchAll();
+    $output = '';
+    foreach ($entries as $entry) {
+        $output .= '<i class="' . $entry['value'] . '"></i>';
+    }
+    return $output;
+}
+
+/*
+ * Clamps the project ID to an existing project in the database. If the ID is higher than any existing number,
+ * we loop back around to the first project ID. If it is lower, we loop back around to the last project ID. If
+ * in the event the autoincrement screws up and we have missing numbers we will round to the nearest existing number.
+ *
+ * @param int The target project ID recieved from the GET string.
+ *
+ * @return int The corrected ID of the project we will be displaying.
+ */
+function clampProjectID(PDO $db, int $id, bool $up) {
```

Comment:
you shouldnt need to do any of the ID manipulation at all. if this is required, it shows there is a deeper problem with your code. how could you ever have an ID that is not in the db. and why would you not just display an error if you did.

Reply:
My reasoning was that if you add a new project, then delete the project in the middle, moving the next project will attempt to load the middle project, but that project does not exist anymore in the db so it won't load anything. This was my attempt to fix that.

Reply:
attempted rework in 3fbee2b

this probably isn't much better but it's late and my brain hurts